vm/vm.go

```diff
@@ -487,18 +526,63 @@ func unlockOutput(vmParams *Params, output iotago.Output, inputIndex uint16) err
 				return ierrors.Wrapf(iotago.ErrInvalidInputUnlock, "input %d's address is already unlocked through input %d's unlock but the input uses a non referential unlock", inputIndex, unlockedAtIndex)
 			}
 
-			if err = vmParams.WorkingSet.UnlockedIdents.SigUnlock(owner, vmParams.WorkingSet.EssenceMsgToSign, uBlock.Signature, inputIndex); err != nil {
+			if err := vmParams.WorkingSet.UnlockedIdents.SigUnlock(owner, vmParams.WorkingSet.EssenceMsgToSign, uBlock.Signature, inputIndex, checkUnlockOnly); err != nil {
 				return ierrors.Join(iotago.ErrUnlockBlockSignatureInvalid, err)
 			}
 
+		default:
+			return ierrors.Wrapf(iotago.ErrInvalidInputUnlock, "input %d has a direct unlockable address (%T) but its corresponding unlock is of type %T", inputIndex, owner, unlock)
+		}
+
+	case *iotago.MultiAddress:
+		switch uBlock := unlock.(type) {
+		case iotago.ReferentialUnlock:
+			if uBlock.Chainable() || !uBlock.SourceAllowed(ownerIdent) {
+				return ierrors.Wrapf(iotago.ErrInvalidInputUnlock, "input %d has none chain address of %s but its corresponding unlock is of type %s", inputIndex, owner.Type(), unlock.Type())
```

Comment:
```suggestion
				return ierrors.Wrapf(iotago.ErrInvalidInputUnlock, "input %d is a chain address of type %s but its corresponding unlock is of type %s", inputIndex, owner.Type(), unlock.Type())
```

I know this error message was copied but it's hard to understand. Not sure if this is much better, but at least a bit.

If I understand correctly, this is the case when I use a chain address and match it with an incorrect unlock, e.g. I try to unlock an NFT Address with an Account Unlock or an Account Address with an NFT Unlock.

If you accept this suggestion, also update the other occurence.